### CS/ECE/ISyE 524 &mdash; Introduction to Optimization &mdash; Spring 2021 ###

### Final Course Project: Due 5/2/21, 12:05pm

# UW-Madison Underguade Course Planning #

#### Shawn Zhong (shawn.zhong@wisc.edu), Evan Wang (xwang2488@wisc.edu), Jun Lin (tan65@wisc.edu)

*****

### Table of Contents

<span style="color:red">TODO: update this</span>

1. [Introduction](#1.-Introduction)
  1. [Constraints](#1.A-Constraints)
  1. [Data](#1.B-Data)
  1. [Outline](#1.C-Outline)

1. [Mathematical Model](#2.-Mathematical-Model)
  1. [Assumptions](#2.A-Assumptions)
  1. [Notations](#2.B-Notations)
  1. [Decision Variables](#2.C-Decision-Variables)
  1. [Constraints](#2.E-Constraints)
  1. [Standard Form](#2.F-Standard-Form)

1. [Solution](#3.-Solution)

1. [Results and Discussion](#4.-Results-and-discussion)
  1. [Optional Subsection](#4.A.-Feel-free-to-add-subsections)

1. [Conclusion](#5.-Conclusion)

## 1. Introduction

The topic that our team settled on is undergrad course planning at UW-Madison. Selecting the appropriate courses for every semester can be a challenging process for some students who aren't certain about the degree structure and numerous requirements. 

Given that students picked their courses for every semester without considering future courses that have strict prerequisites, some of which are compulsory courses, it is very likely that the student will pick a non-optimum route for graduation. For example, a student will not be able to enroll in a compulsory course in a later semester if he/she hasn't taken the prerequisites to that course. As a result, he/she will have to enroll in the prerequisite courses before being able to enroll in the compulsory courses in subsequent semesters, which will delay the student's graduation date. 

The goal of this project is to find an optimal course selection for each semester in regards to completing the graduation requirements within the shortest period and the least number of courses taken.

### 1.A Constraints

We consider the following constraints for course planning:

1. **Prerequisites**: Some courses are required to be completed before enrolling in certain classes

2. **Graduation requirements**: Some courses are required to be taken to meet the graduation requirements

3. **Credit Limit**: The number of courses taken in any given semester cannot exceed a certain number of credit hours

4. **No retaking**: A course cannot be taken twice


We also have the following constraints that can be added interactively: 

5. **Desired courses**: Through the interactive console, the student can specify which courses must be included in the plan. 

6. **Limit workload**: We also allow the student to cap the maximum credit. 

7. **Transferred courses**: We allow user to specify the transferred courses before entering the university, so our planning program can take that into account. 

### 1.B Data

- The graduation requirements are gathered from the [Online Underguade Guide](https://guide.wisc.edu/undergraduate/#majorscertificatestext). 

- The course list and prerequisite relationships are obtained from the [Course Search and Enroll website](https://public.enroll.wisc.edu/).


### 1.C Outline

The rest of the report is structured as follows: We first introduce the mathematical model for this problem ... <span style="color:red">TODO: update this</span>

## 2. Mathematical Model

### 2.A Assumptions

We made the following assumptions for our course planning problem: 

1. We assume that there is no time conflict for class attendance in a given semester. One can think that all the courses can be taken asynchronously in this pandemic time. This assumption is made since we haven't figured out a way to gather section-level data, and it complicates the optimization problem. 

1. We assume that the student can pass all the courses taken so there is no retaking. 

<!-- 1. For graduation requirements, we only consider the major requirements. Students can limit the maximum credit per semester to make space for the courses satisfying general education requirements.  -->

### 2.B Notations


| Variable | Description | Example  |
| :---: | :---: | :---: |
| $T$ | maximum number of semesters | For 4-year college, we have $T = 8$ |
| $$t\in \{0, \cdots, T\}$$ | a specific semester      | $t = 2$ is the second semester <br /> $t = 0$ is used for transferred courses |
| $C$ | the set of all classes      |   $$C = \{ \text{CS 524}, \text{MATH 240}, \cdots \}$$|
| $$c \in C$$ | a specific class | $$c = \text{CS 524}$$ |

### 2.C Decision Variables

$x[t, c] \in \{ 0, 1\}$ is a Boolean variable to denote whether to take the class $c \in C$ at semester $t \in \{0, \cdots, T\}$. 

For example, if a student takes CS 524 on the thrid semester, then $x[3, \text{CS 524}]= 1$. If the student has CS 200 transferred, then $x[0, \text{CS 200}]= 1$

### 2.D Objective 

A naive objective would be to minimize the sum of $x$: 

$$\min \sum_{t=0}^{T} \sum_{c \in C} x[t, c]$$

But this would lead to many equally good solutions. To avoid this, we add a weight $t$ to the class $c$ if its taken at semester $t$

$$\min \sum_{t=0}^{T} \sum_{c \in C} x[t, c] \cdot t$$

This avoid students procrastinating class to later semesters. 

### 2.E Constraints

#### Prerequisite

To take the class $c$, students may need to take $c'$ in the previous semesters, we can encode such prerequisite using the following constraint: 

$$x[t, c] \le \sum_{i=0}^{t - 1} x[i, c'] $$


For example, CS 524 has the prerequisite similar to "(CS 200 or 300) and (MATH 340 or 341)", we can encode it with the constraint:

$$x[t, \text{CS 524}] \le \sum_{i=0}^{t - 1} x[i, \text{CS 200}] + \sum_{i=0}^{t - 1} x[i, \text{CS 300}] $$

$$x[t, \text{CS 524}] \le \sum_{i=0}^{t - 1} x[i, \text{CS 340}] + \sum_{i=0}^{t - 1} x[i, \text{CS 341}] $$


#### Graduation Requirement

In order to meet the graduation requirement, some class $c$ must be taken, so we have the constraint:

$$\sum_{t=0}^{T} x[t, c] \ge 1$$
    
There are other kinds of graduation requirements, such as taking $k$ coureses from a list of $C$, we can encode such requirements using the following constraint:

$$\sum_{t=0}^{T} \sum_{c \in C} x[t, c] \ge k$$
    
#### Max credits

For undergraduate, the number of credits taken in a given semester cannot exceeds a certain number, so we need to ensure that

$$\sum_{c \in C} x[t, c] \cdot \text{credit}(c) \le \text{max_credit}, \quad \text{for all semesters } t \in \{1, \cdots, T\}, \text{classes } c \in C$$

#### No retaking


$$\sum_{t=1}^{T} x[t, c] \le 1, \quad \text{for all classes } c \in C$$

### User specified parameters through interractive interface

#### Transferred Courses

$$ x[0, c] = 1, \quad \text{for all classes } c \in C_{\text{trans}}$$

$$ x[0, c] = 0, \quad \text{for all classes } c \in C \setminus C_{\text{trans}}$$


#### Desired Courses

$$\sum_{t=1}^{T} x[t, c] \ge 1 \quad \text{for all classes } c \in C_{\text{desiredCourses}} $$

#### Limit workload

The user will be able to set the desired maximum workload every semester, between 3 and 22, 

$$\sum_{c \in C} x[t, c] \cdot \text{credit}(c) \le \text{maxWorkload}, \quad \text{for all semesters } t \in \{1, \cdots, T\}$$

### 2.F Standard Form

We model this problem as a mixed integer programming problem, and the standard form is shown below

$$
\begin{aligned}
\underset{x}{\text{minimize}}
\quad& \sum_{t=0}^{T} \sum_{c \in C} x[t, c] \cdot t \\
\text{subject to:}
\quad&
x[t, c] \le \sum_{i=0}^{t - 1} x[i, c'], \\
    & \qquad
    \text{for all classes } c \in C, 
    \text{classes } c' \in \text{prerequisite}(c), 
    \text{semesters } t \in \{1, \cdots, T\} \\
\quad&
\sum_{t=0}^{T} \sum_{c \in C_i} x[t, c] \ge k_i \\
    & \qquad
    \text{for all requirement set } C_i, \text{required number of courses } k_i \\
\quad&
\sum_{t=0}^{T} x[t, c] \le 1, \\
    & \qquad \text{for all class } c \in C \\
\quad&
\sum_{c \in C} x[t, c] \cdot \text{credit}(c) \le \text{max_credit}, \\
    & \qquad
    \text{for all semesters } t \in \{1, \cdots, T\}, \text{classes } c \in C \\
\quad& 
  x[0, c] = 1, \quad \text{for all classes } c \in C_{\text{trans}}\\
\quad& 
x[0, c] = 0, \quad \text{for all classes } c \in C \setminus C_{\text{trans}} \\
\quad& x[t, c] \in \{0, 1\}, \\
    & \qquad
    \text{for all semesters } t \in \{0, \cdots, T\}, \text{classes } c \in C 
\end{aligned}
$$

## 3. Solution

### 3.A Data Gathering

We first collect the course data via the public facing API from from the [Course Search and Enroll website](https://public.enroll.wisc.edu/). 

For example, the following URL gives information about the courses provided in Spring 2021 (with term code 1214). 

https://public.enroll.wisc.edu/api/search/v1?query={"selectedTerm":"1214"}

We wrote a short script to download the data, and saved the file to `data/1214-spring-2021.json`, so that it can be later procseed by Julia. 

### 3.B Data Preprossing

We first use the `JSON` package to load the raw data into a Julia list. 

In [1]:
using Pkg
Pkg.add("JSON")

   Updating registry at `~/.julia/registries/General`
  Resolving package versions...
No Changes to `~/.julia/environments/v1.5/Project.toml`
No Changes to `~/.julia/environments/v1.5/Manifest.toml`


In [2]:
using JSON

raw_data = JSON.parsefile("data/1214-spring-2021.json")["hits"];
print("There are ", length(raw_data), " items in raw_data")

There are 6510 items in raw_data

Each item in the `raw_data` array is a dictionary with the keys listed below. An example item in the array can be found in the [Appendix](#Appendix). 

In [3]:
join(keys(raw_data[1]), ", ")

"honors, allCrossListedSubjects, breadths, matched_queries, termCode, levels, subjectAggregate, courseId, academicGroupCode, gradingBasis, advisoryPrerequisites, ethnicStudies, lettersAndScienceCredits, approvedForTopics, courseDesignationRaw, openToFirstYear, gradCourseWork, catalogPrintFlag, sustainability, instructorProvidedContent, courseRequirements, subject, repeatable, fullCourseDesignationRaw, typicallyOffered, foreignLanguage, enrollmentPrerequisites, titleSuggest, minimumCredits, title, lastUpdated, workplaceExperience, courseDesignation, generalEd, topics, description, lastTaught, creditRange, catalogSort, currentlyTaught, firstTaught, catalogNumber, maximumCredits, fullCourseDesignation"

We notice that the same course can appear multiple times in the list `raw_data` since a course can be cross-listed in multiple departments, so we need to add another layer of indirection to use the course id instead of course name as the unique identifier for a given course. 

The mapping between from the course name to the course id is saved in the dictionary `cls_name_to_id`, and the mapping from the id to the actual data is saved in `cls_dict`. 

In [4]:
cls_dict = Dict(
    Symbol(c["courseId"]) => c 
    for c in raw_data
)

cls_name_to_id = Dict(
    c["courseDesignation"] => Symbol(c["courseId"])
    for c in raw_data 
)

for cls_name in ["COMP SCI 525", "I SY E 525", "MATH 525", "STAT 525"]
    println("The course id for \"", cls_name, "\" is ", cls_name_to_id[cls_name])
end

The course id for "COMP SCI 525" is 004272
The course id for "I SY E 525" is 004272
The course id for "MATH 525" is 004272
The course id for "STAT 525" is 004272


In [5]:
println("We have ", length(cls_dict), " courses in total")

We have 5611 courses in total


### 3.C Decision Variables

The decision variable is a Boolean matrix of shape $5611 \times 8$. 

In [6]:
using JuMP, Gurobi

C = keys(cls_dict)
T = 0:8

m = Model(with_optimizer(Gurobi.Optimizer, LogToConsole=0))
@variable(m, x[C, T], Bin)
size(x)



--------------------------------------------
--------------------------------------------

Academic license - for non-commercial use only - expires 2021-05-10


(5611, 9)

### 3.D CS Major Requirements

We use the CS major requirements from the following URL to construct the constraints. 

https://guide.wisc.edu/undergraduate/letters-science/computer-sciences/computer-sciences-ba/#requirementstext. 

We note that all of the CS constraints are in the form of choosing $k$ courses from a course set $C'$.  

In [7]:
"""
Convert CS course number to course id
"""
get_cs_cls_id(number) = get(cls_name_to_id, "COMP SCI " * string(number), Symbol())

"""
Convert a list of CS course numbers to a list of course ids
"""
cs_ids(arr...) = filter(id -> id in C, [get_cs_cls_id(e) for e in arr])


function add_cs_major_req()
    CS_basic = cs_ids(240, 252, 300, 354, 400)
    CS_theory = cs_ids(577, 520)
    CS_xware = cs_ids(
        407, 506, 536, 538, 537, 552, 564, 640, 642
    )
    CS_app = cs_ids(
        412, 425, 513, 514, 524, 525, 534, 540, 545, 547, 559, 570
    )
    CS_elec = cs_ids(
        407, 412, 425, 435, 471, 475, 506, 513, 514, 520, 524, 525, 526, 532, 
        533, 534, 536, 537, 538, 539, 540, 545, 547, 552, 558, 559, 564, 567, 
        570, 576, 577, 579, 635, 640, 642, 679, 639
    )

    
    # Take all from basic computer sciences
    for c in CS_basic
        @constraint(m, sum(x[c, t] for t in T) >= 1)
    end

    # Complete 1 for Theory of computer science
    @constraint(m, sum(x[c, t] for t in T for c in CS_theory) >= 1)

    # Complete 2 for Software & Hardware
    @constraint(m, sum(x[c, t] for t in T for c in CS_xware) >= 2)

    # Complete 1 for Applications
    @constraint(m, sum(x[c, t] for t in T for c in CS_app) >= 1)

    # Complete 2 for Electives
    @constraint(m, sum(x[c, t] for t in T for c in CS_elec) >= 2)
end;

### 3.E Math Major Requirement

We use the Math major requirements from the following URL to construct the constraints. 

https://guide.wisc.edu/undergraduate/letters-science/mathematics/mathematics-ba/mathematics-mathematics-programming-computing-ba/#requirementstext

The constraints for Math major are more complicated, which includes the following forms:

- Choosing $k$ courses from a course set $C'$

- $c_1$ and $c_2$ can be either both taken to count as satisfying one course in a course set $C'$

- Only one of the course $c \in C'$ should be taken. If $c_1 \in C'$ has been taken before, then the student shouldn't take $c_2 \in C'$. 

In [8]:
"""
Convert Math course number to course id
"""
get_math_cls_id(number) = get(cls_name_to_id, "MATH " * string(number), Symbol())

"""
Convert a list of Math course numbers to a list of course ids
"""
math_ids(arr...) = filter(id -> id in C, [get_math_cls_id(e) for e in arr])

function add_math_major_req()
    MATH_linear_algebra = math_ids(320, 341, 340, 375)
    MATH_intermediate = math_ids(321, 322, 375, 421, 467)
    MATH_advanced = math_ids(514, 521, 531, 535, 540, 541, 571)
    MATH_elective_A = math_ids(
        513, 522, 525, 542, 567, 570,
        605, 619, 627, 629, 632, 635
    )
    MATH_elective_B = math_ids(
        310, 319, 376, 415, 425, 431, 309, 435, 443, 475
    )
    CS_basic = cs_ids(300, 400)
    CS_elective = cs_ids(
        412, 471, 520, 524, 526, 532, 533, 534, 538, 539,
        540, 545, 558, 559, 567, 576, 577, 635, 642
    )
    
    # Complete 1 for Linear Algebra
    @constraint(m, sum(x[c, t] for t in T for c in MATH_linear_algebra) >= 1)
    
    # Complete 1 for Linear Algebra Intermediate Mathematics Requirement
    @constraint(m, sum(x[c, t] for t in T for c in MATH_intermediate) >= 1)
    
    # Complete 1 for Advanced Mathematics Requirement
    @constraint(m, sum(x[c, t] for t in T for c in MATH_advanced) >= 1)
    
    # MATH Elective to reach required 6 courses
    # Select one or more from MATH_elective_A
    @constraint(m, sum(x[c, t] for t in T for c in MATH_elective_A) >= 1)
    # Select Select remaining courses from MATH_elective_B
    @constraint(m, 
        sum(x[c, t] for t in T for c in MATH_elective_A) +
        sum(x[c, t] for t in T for c in MATH_elective_B) >= 6
    )
    
    # Programming and Computations Requirement at least 12 credit hours
    for c in CS_basic
        @constraint(m, sum(x[c, t] for t in T) >= 1)
    end
    # CS_elective
    @constraint(m, sum(x[c, t] for t in T for c in CS_elective) >= 2)
    
    # Additional requirements
    # 321 & 322 must be taken together
    @constraint(m, 
        sum(x[get_math_cls_id(321), t] for t in T) == 
        sum(x[get_math_cls_id(322), t] for t in T)
    )
    
    # 319, 320, & 376 cannot be taken together
    @constraint(m, 
        sum(x[c, t] for t in T for c in math_ids(319, 320, 376)) <= 1
    )
    
    # 309 & 431 cannot be taken together
    @constraint(m, 
        sum(x[c, t] for t in T for c in math_ids(309, 431)) <= 1
    )
end;

### 3.F Course Prerequisites

Although we have the course prerequisites information in `raw_data`, it's in string format (see [Appendix](#6.B-Enrollment-Prerequisites)), so we need to manually encode this information. 

<span style="color:red">
JUN LIN CHANGED 1:t-1 to 0:t-1, because prerequisites accounts for all previous semester including transferred
</span>

In [9]:
"""
A helper function for adding prerequisites

Don't use this function directly. 
"""
function _add_prereq(id1, id2, cls, prereq)
    for t in T
        prereq_id = filter(id -> id in C, [id2(p) for p in prereq])
        @constraint(m, x[id1(cls), t] <= sum(x[id, i] for i in 0:t-1 for id in prereq_id))
    end
end

"""
Specify CS class prerequisite for a given CS class
"""
add_cs_cs_prereq(cs_cls, one_of...) = 
    _add_prereq(get_cs_cls_id, get_cs_cls_id, cs_cls, one_of)

"""
Specify Math class prerequisite for a given CS class
"""
add_cs_math_prereq(cs_cls, one_of...) =
    _add_prereq(get_cs_cls_id, get_math_cls_id, cs_cls, one_of)

"""
Specify Math class prerequisite for a given Math class
"""
add_math_math_prereq(math_cls, one_of...) =
    _add_prereq(get_math_cls_id, get_math_cls_id, math_cls, one_of)


"""
Add all course prerequisite
"""
function add_all_prereq()
    add_cs_cs_prereq(300, 200)
    add_cs_cs_prereq(354, 252)
    add_cs_cs_prereq(354, 300)

    add_cs_cs_prereq(506, 400)
    add_cs_cs_prereq(506, 407, 536, 537, 559, 564, 570, 679, 552)

    add_cs_cs_prereq(552, 352)
    add_cs_cs_prereq(552, 354)
    
    add_cs_cs_prereq(559, 400)

    for c in [400, 407, 412, 513, 514, 524, 532, 534, 539, 540, 570, 576]
        add_cs_cs_prereq(c, 300)
    end

    for c in [536, 537, 538, 558, 564]
        add_cs_cs_prereq(c, 354)
        add_cs_cs_prereq(c, 400)
    end

    for c in [520, 577]
        add_cs_cs_prereq(c, 400)
        add_cs_cs_prereq(c, 240, 475)
    end

    for c in [640, 642]
        add_cs_cs_prereq(c, 537)
    end
    
    for c in [435, 513, 524, 525]
        add_cs_math_prereq(c, 340, 341, 375)
    end

    for c in [412, 532, 576]
        add_cs_math_prereq(c, 222)
    end
    
    for c in [425, 475, 513, 533, 567]
        add_cs_math_prereq(c, 320, 340, 341, 375)
    end
    
    add_cs_math_prereq(412, 240, 234)
    add_cs_math_prereq(435, 320)
    add_cs_math_prereq(558, 234)
    
    
    
    
    for c in [234, 310, 319, 320, 340]
        add_math_math_prereq(c, 222, 276)
    end
    
    for c in [309, 321, 341, 421, 431]
        add_math_math_prereq(c, 234, 376)
    end
    
    for c in [321, 443, 540]
       add_math_math_prereq(c, 319, 320, 340, 341, 375) 
    end
    
    
    add_math_math_prereq(222, 221)
    add_math_math_prereq(322, 321)
    add_math_math_prereq(376, 375)
    
    
    add_math_math_prereq(521, 234, 322, 341, 376, 421)
    add_math_math_prereq(522, 521)
    
    add_math_math_prereq(531, 376, 421, 521)
    
    add_math_math_prereq(540, 234, 375)
    add_math_math_prereq(540, 341, 375, 421, 476, 521)
    
    add_math_math_prereq(541, 234, 375)
    add_math_math_prereq(541, 320, 341, 375, 421, 476, 521)
    
    add_math_math_prereq(542, 541)
    add_math_math_prereq(567, 541)
    
    add_math_math_prereq(619, 322, 421, 521)
    add_math_math_prereq(619, 319, 320, 376, 415, 519)
    
    add_math_math_prereq(629, 552)
    
    add_math_math_prereq(632, 431, 309, 531)
    add_math_math_prereq(632, 320, 340, 341, 375, 421, 531)
end;

# x = find_optimal_schedule(max_credit=9, math_major=true)
# print_result(x)

### 3.G Other Constraints

We define the max credit constraint and no retaking constraint in the following block: 


In [10]:
"""
Returns the minimum number of credit given a course id
"""
credit(c) = cls_dict[c]["minimumCredits"]

function add_max_credit_constraint(max_credit)
    for t in 1:length(max_credit)
        @constraint(m, sum(x[c, t] * credit(c) for c in C) <= max_credit[t])
    end
end;

function add_no_retaking_constraint()
    for c in C
        @constraint(m, sum(x[c, t] for t in T) <= 1)
    end
end;

function add_transferred_courses_constraint(transferred_courses)
    for c in C
        if (c in transferred_courses)
            @constraint(m, x[c, 0] == 1)
        else
            @constraint(m, x[c, 0] == 0)
        end
    end
end;

function add_desired_courses_constraint(desired_courses)
    for c in desired_courses
        @constraint(m, sum(x[c, t] for t in T) >= 1)
    end
    
end;

### 3.H Putting Things Together

We define the `find_optimal_schedule` function below to solve for the optimal schedule given the constraints above

In [11]:
function find_optimal_schedule(;
        num_semester=8,
        max_credit=[], 
        math_major=false, 
        cs_major=false,
        transferred_courses=[],
        desired_courses=[]
    )
    global T = 0:num_semester
    global m = Model(with_optimizer(Gurobi.Optimizer, LogToConsole=0))
    @variable(m, x[C, T], Bin)
    
    add_all_prereq()
    add_max_credit_constraint(max_credit)
    add_no_retaking_constraint()
    add_transferred_courses_constraint(transferred_courses)
    add_desired_courses_constraint(desired_courses)
    
    if math_major
        add_math_major_req()
    end
    
    if cs_major
        add_cs_major_req()
    end
    
    @objective(m, Min, sum(x[c, t] * t for c in C, t in 1:num_semester))

    @time optimize!(m)
    if termination_status(m) == MOI.OPTIMAL
        global x = value.(x)
        return x
    else
        println(termination_status(m))
    end
end

find_optimal_schedule();


--------------------------------------------
--------------------------------------------

Academic license - for non-commercial use only - expires 2021-05-10
  5.720953 seconds (30.32 M allocations: 1.546 GiB, 7.29% gc time)


## 4. Results and Discussion ##

Here, you display and discuss the results. Show figures, plots, images, trade-off curves, or whatever else you can think of to best illustrate your results. The discussion should explain what the results mean, and how to interpret them. You should also explain the limitations of your approach/model and how sensitive your results are to the assumptions you made.

### 4.A Show Schedule

We first define some functions to print out the results from the optimization problem. 

In [12]:
"""
Returns the name for a course given the course id
"""
function get_cls_name(id)
    cls = cls_dict[id]
    subjects = [
        e["shortDescription"] 
        for e in cls["allCrossListedSubjects"]
    ]
    if length(subjects) == 0
        return cls["courseDesignation"]
    else
        return join(subjects, "/") * " " * cls["catalogNumber"]
    end
end

"""
Returns the full name for a course given the course id
"""
get_cls_full_name(id) = get_cls_name(id) * ": " * cls_dict[id]["title"]

"""
Print the schedule
"""
function print_result(x)
    n_cls = 0
    total_credit = 0
    for t in T
        for c in C
            if x[c, t] > 0
                name = get_cls_full_name(c)
                if(t == 0)
                    println("Transfered: \"", name, "\"")
                else
                    println("Semester ", t, ": take \"", name, "\"")
                    n_cls += 1
                    total_credit += credit(c)
                end                
                
            end
        end
    end
    println("Number of courses taken: ", n_cls, " with ", total_credit, " credits")
    return n_cls, total_credit
end;

### 4.B CS Major

In [13]:
num_semester = 8
max_credit = [6 for i=1:num_semester]
x = find_optimal_schedule(max_credit = max_credit, cs_major=true, num_semester = num_semester)
print_result(x);


--------------------------------------------
--------------------------------------------

Academic license - for non-commercial use only - expires 2021-05-10
  0.449329 seconds (1.55 M allocations: 109.690 MiB, 4.34% gc time)
Semester 1: take "COMP SCI/MATH 240: Introduction to Discrete Mathematics"
Semester 1: take "COMP SCI 200: Programming I"
Semester 2: take "COMP SCI/E C E 252: Introduction to Computer Engineering"
Semester 2: take "COMP SCI 300: Programming II"
Semester 3: take "COMP SCI/E C E 354: Machine Organization and Programming"
Semester 3: take "COMP SCI 400: Programming III"
Semester 4: take "COMP SCI 520: Introduction to Theory of Computing"
Semester 4: take "COMP SCI 538: Introduction to the Theory and Design of Programming Languages"
Semester 5: take "COMP SCI 536: Introduction to Programming Languages and Compilers"
Semester 5: take "COMP SCI 534: Computational Photography"
Number of courses taken: 10 with 29 credits


### 4.C Math Major

Unlike the CS major, we are unable to find a solution if we only take 6 credits per semester

In [14]:
num_semester = 8
max_credit = [6 for i=1:num_semester]
x = find_optimal_schedule(max_credit = max_credit, math_major=true, num_semester = num_semester)
print_result(x);


--------------------------------------------
--------------------------------------------

Academic license - for non-commercial use only - expires 2021-05-10
  0.119062 seconds (183.50 k allocations: 39.087 MiB)
INFEASIBLE_OR_UNBOUNDED


LoadError: MethodError: no method matching getindex(::Nothing, ::Symbol, ::Int64)

In [15]:
num_semester = 8
max_credit = [9 for i=1:num_semester]
x = find_optimal_schedule(max_credit = max_credit, math_major=true, num_semester = num_semester)
print_result(x);


--------------------------------------------
--------------------------------------------

Academic license - for non-commercial use only - expires 2021-05-10
  0.137730 seconds (183.52 k allocations: 39.087 MiB)
Semester 1: take "COMP SCI 200: Programming I"
Semester 1: take "MATH 221: Calculus and Analytic Geometry 1"
Semester 2: take "MATH 222: Calculus and Analytic Geometry 2"
Semester 2: take "COMP SCI 300: Programming II"
Semester 3: take "COMP SCI/MATH 514: Numerical Analysis"
Semester 3: take "MATH 340: Elementary Matrix and Linear Algebra"
Semester 3: take "COMP SCI 400: Programming III"
Semester 4: take "COMP SCI/I SY E/MATH 425: Introduction to Combinatorial Optimization"
Semester 4: take "MATH 319: Techniques in Ordinary Differential Equations"
Semester 4: take "MATH 443: Applied Linear Algebra"
Semester 5: take "COMP SCI 540: Introduction to Artificial Intelligence"
Semester 5: take "COMP SCI/E C E 533: Image Processing"
Semester 5: take "COMP SCI/I SY E/MATH/STAT 525: Li

### 4.D CS & Math Double Major

With double major, only 21 classes taken instead of 17 + 10 = 27, and the number of credits is reduced as well.  

In [16]:

num_semester = 8
max_credit = [9 for i=1:num_semester]
x = find_optimal_schedule(max_credit = max_credit, cs_major=true, math_major=true, num_semester = num_semester)
print_result(x);


--------------------------------------------
--------------------------------------------

Academic license - for non-commercial use only - expires 2021-05-10
  0.148213 seconds (183.68 k allocations: 39.119 MiB)
Semester 1: take "COMP SCI 200: Programming I"
Semester 1: take "MATH 221: Calculus and Analytic Geometry 1"
Semester 2: take "MATH 222: Calculus and Analytic Geometry 2"
Semester 2: take "COMP SCI/E C E 252: Introduction to Computer Engineering"
Semester 2: take "COMP SCI 300: Programming II"
Semester 3: take "COMP SCI/E C E 354: Machine Organization and Programming"
Semester 3: take "MATH 340: Elementary Matrix and Linear Algebra"
Semester 3: take "COMP SCI 400: Programming III"
Semester 4: take "COMP SCI/I SY E/MATH 425: Introduction to Combinatorial Optimization"
Semester 4: take "COMP SCI/MATH 240: Introduction to Discrete Mathematics"
Semester 4: take "COMP SCI/MATH 514: Numerical Analysis"
Semester 5: take "COMP SCI 536: Introduction to Programming Languages and Compil

### 4.E Interactive console interface

##### All user input are passed through a robust error checking script to check for invalid inputs, which will cause difficult to trace errors later on if not properly dealt with

#### Desired courses
User will be able to choose which are the courses they want to be included in the course schedule. Note that all the prerequisites of the desired courses will be accounted for as well. 

#### Transferred courses
User will be able to specify which are the courses they already completed prior to enrolling into the university. Note that only courses without any prerequisites will be allowed. This is based on the assumption that transfered courses are elementary level courses. 

#### Max work load on each semester
User is able to specify the maximum number of credits they want to take on a given semester between 0 and 22. Note that maximum credit load of "0" would mean the user wants to take a gap semester 

#### Change number of semester
User is able to modify the total number of semesters. The default number of semester is 8. Changes on the maximum credit load each semester made through "Max work load on each semester" will be preserved. If the new specified number of semester is greater than the current value, the maximum credit loads of additional semester will be set to the default number, 9. 

In [39]:
desired_courses = []
desired_course_ids = []
transferred_courses = []
transferred_course_ids = []
num_semester = 8
max_work_load = [9 for i in 1: num_semester]


function responseOne(desired_courses)
    print("format:\n eg: \"COMP SCI 300\"\n")
    courseName = readline()
    
    if(!haskey(cls_name_to_id, courseName))
        println("Invalid course name/ Course not found")
        return desired_courses
    end
    
    println(courseName, " is successfully added to desired course list\n ")
    return vcat(desired_courses, courseName)
end


function responseTwo(transferred_courses)
    print("format:\n eg: \"COMP SCI 300\"\n")
    
    courseName = readline()
    
    if(!haskey(cls_name_to_id, courseName))
        println("Invalid course name/ Course not found")
        return transferred_courses
    end
    
    println(courseName, " is successfully added to transfered courses list\n ")
    return vcat(transferred_courses, courseName)
end

function responseThree(max_work_load)

    println("Enter maximum Y credits for X semester, default workload is 9 credits per semester")
    println("format:\n eg: \"2, 6\" means maximum of 6 credits in semester 2")
    
    response = readline()
    
    if(size(split(response, ','))[1] != 2)
        print("Invalid format")
        return max_work_load
    end
    
    semester = split(response, ',')[1]
    max_credit = split(response, ',')[2]
    

    re = r"^([1-9])";
    
    if(!occursin(re, semester))
        println("Invalid semester number format: \n")
        return max_work_load
    end
    
    semester = parse(Int64, semester)
    if(semester > num_semester)
        println("Invalid semester number: semester input is greater than number of semester")
        return max_work_load
    end

    # remove space at the beginning if the user input a space
    if(max_credit[1] == ' ')
        max_credit = max_credit[2:sizeof(max_credit)]
    end
    
    re = r"^([0-9])";
    if(!occursin(re, max_credit))
        println("Invalid max credit number format: \n")
        return max_work_load
    end

    max_credit = parse(Int64, max_credit)

    if ((max_credit < 0) || (max_credit > 22 ))
        println("Invalid credit amount, please enter a number between 0 and 22")
        return max_work_load
    end
    
    max_work_load[semester] = max_credit
    println("Maximum workload on semester: ", semester," successfully set to: ", max_credit)
    
    return max_work_load
end

function responseFour(num_semester)

    println("Enter number of semester")
    
    response = readline()

    re = r"^([0-9])";
    if(!occursin(re, response))
        println("Invalid number, number of semester is still: ", num_semester)
        return num_semester
    end
    
    response = parse(Int64, response)

    
    if ((response < 4))
        println("You're not a genius, lets be real here. number of semester is still: ", num_semester)
        return num_semester
    end
    
    if ((response > 12))
        println("You're not seriosly gonna spend more than 6 years in undergrad are you? number of semester is still: ", num_semester)
        return num_semester
    end
    
    println("number of semester successfully set to: ", response)
    
    return response
end

"""
Convert a single course name to course ID
"""
get_cls_id(cls_name) = get(cls_name_to_id, cls_name, Symbol())

"""
Convert a list of course names to a list of course ids
"""
get_course_ids(arr) = filter(id -> id in C, [get_cls_id(e) for e in arr])

loopV = true

while(loopV)
    
    print("\nWhat would you like to do? (enter number)\n
        1. Add desired courses
        2. Add prior transferred courses 
        3. Limit workload on each semester
        4. Change number of semester, default is 8
        5. Run program
        \n") 
  
    # Calling rdeadline() function
    response = readline()
    
    if (response == "1")
        desired_courses = responseOne(desired_courses)
        
    elseif(response == "2")
        transferred_courses = responseTwo(transferred_courses)
        
    elseif(response == "3")
        max_work_load = responseThree(max_work_load)
        
    elseif(response == "4")
        num_semester = responseFour(num_semester)
        if(length(max_work_load) > num_semester)
            max_work_load = max_work_load[1: num_semester]
        else
            max_work_load = vcat(max_work_load, [9 for i in 1: num_semester - length(max_work_load)])
        end
        
        println("max_work_load after changing number of semester: ", max_work_load)
        
    elseif(response == "5")
        desired_course_ids = get_course_ids(desired_courses)
        transferred_course_ids = get_course_ids(transferred_courses)
        loopV = false
        
        # convert list of String into the correct format
    else
        print("invalid response \n")
    end
    
end

x = find_optimal_schedule(max_credit = max_work_load, cs_major=true, 
    desired_courses = desired_course_ids, 
    transferred_courses = transferred_course_ids, num_semester = num_semester)
print_result(x);


What would you like to do? (enter number)

        1. Add desired courses
        2. Add prior transferred courses 
        3. Limit workload on each semester
        4. Change number of semester, default is 8
        5. Run program
        
stdin> 1
format:
 eg: "COMP SCI 300"
stdin> MUSIC 113
MUSIC 113 is successfully added to desired course list
 

What would you like to do? (enter number)

        1. Add desired courses
        2. Add prior transferred courses 
        3. Limit workload on each semester
        4. Change number of semester, default is 8
        5. Run program
        
stdin> 1
format:
 eg: "COMP SCI 300"
stdin> COMP SCI 240
COMP SCI 240 is successfully added to desired course list
 

What would you like to do? (enter number)

        1. Add desired courses
        2. Add prior transferred courses 
        3. Limit workload on each semester
        4. Change number of semester, default is 8
        5. Run program
        
stdin> 1
format:
 eg: "COMP SCI 300"
stdin>

In [30]:
test = [9 for i=1:5]
test2 = vcat(test, [7 for i in 1: 1-1])



5-element Array{Int64,1}:
 9
 9
 9
 9
 9

In [19]:
Semester 1: take "COMP SCI/MATH 240: Introduction to Discrete Mathematics"
Semester 1: take "COMP SCI 200: Programming I"
Semester 1: take "COMP SCI/E C E 252: Introduction to Computer Engineering"
Semester 2: take "COMP SCI 300: Programming II"
Semester 3: take "COMP SCI/E C E 354: Machine Organization and Programming"
Semester 3: take "COMP SCI 534: Computational Photography"
Semester 3: take "COMP SCI 400: Programming III"
Semester 3: take "COMP SCI 407: Foundations of Mobile Systems and Applications"
Semester 4: take "COMP SCI/E C E 506: Software Engineering"
Semester 4: take "COMP SCI 520: Introduction to Theory of Computing"
Number of courses taken: 10 with 29 credits

LoadError: syntax: extra token "1" after end of expression

In [20]:
tester = [1, 2, 3, 4, 5, 6, 7, 8, 9]

tester[9]

9

## 5. Conclusion ##

Summarize your findings and your results, and talk about at least one possible future direction; something that might be interesting to pursue as a follow-up to your project.

Future work: some courses is only offered in Spring or Fall, extend to more major <span style="color:red">TODO: update this</span>

## 6. Appendix

### 6.A Course Data Example

In [21]:
show(IOContext(stdout, :limit => false), "text/plain", cls_dict[get_cs_cls_id(524)]);

Dict{String,Any} with 44 entries:
  "honors" => nothing
  "allCrossListedSubjects" => Any[Dict{String,Any}("departmentURI"=>"http://www.cs.wisc.edu/","footnotes"=>Any["Courses taught and managed by the Computer Sciences department often have enrollment restrictions that give students in UW-Madison Computer Sciences programs priority access during initial enrollment periods.\n\nEvening exams are likely for most of our undergraduate courses."],"formalDescription"=>"COMPUTER SCIENCES","undergraduateCatalogURI"=>"http://guide.wisc.edu/undergraduate/letters-science/computer-sciences/","termCode"=>"1214","departmentOwnerAcademicOrgCode"=>"L0780","description"=>"COMPUTER SCIENCES","graduateCatalogURI"=>"http://guide.wisc.edu/graduate/computer-sciences/","uddsFundingSource"=>"A4820","shortDescription"=>"COMP SCI","subjectCode"=>"266","schoolCollege"=>Dict{String,Any}("schoolCollegeURI"=>"http://www.ls.wisc.edu/","shortDescription"=>"Letters and Science","formalDescription"=>"Letters and Scienc

### 6.B Enrollment Prerequisites 

In [22]:
for (k, v) in sort(cls_dict)
    cls_name = get_cls_name(k)
    if occursin("MATH", cls_name)
        printstyled(stdout, cls_name, bold=true)
        println(": ", v["enrollmentPrerequisites"])
    end
end

COMP SCI/I SY E/MATH 425: (MATH 320, 340, 341, or 375) or graduate/professional standing or member of the Pre-Masters Mathematics (Visiting International) Program
COMP SCI/MATH 514: (MATH 320, 340, 341, or 375) and (MATH 322, 376, 421, or 521) and (COMP SCI 200, 220, 300, 310, or 301 prior to Spring 2020) or graduate/professional standing or member of the Pre-Masters Mathematics (Visiting International) Program
COMP SCI/I SY E/MATH/STAT 525: MATH 320, 340, 341, 375, or 443 or graduate/professional standing or member of the Pre-Masters Mathematics (Visiting International) Program
COMP SCI/I SY E/MATH 728: Graduate/professional standing
COMP SCI/I SY E/MATH/STAT 726: Graduate/professional standing
MATH 112: MATH 96 or placement into MATH 112. MATH 118 does not fulfill the requisite
MATH 113: MATH 112 or placement into MATH 113
MATH 114: MATH 96 or placement into MATH 114. MATH 118 does not fulfill the requisite
MATH 130: MATH 96 or placement into MATH 130 and classified as Elementary Edu

MATH 531: MATH 376, 421, or 521 or graduate/professional standing or member of the Pre-Masters Mathematics (Visiting International) Program
CURRIC/MATH 471: (MATH 341, 375, or 421) and MATH 461
MATH 848: Graduate/professional standing or member of the Pre-Masters Mathematics (Visiting International) Program
MATH 849: Graduate/professional standing or member of the Pre-Masters Mathematics (Visiting International) Program
MATH 96: Placement into MATH 96. Department consent required to drop/swap from course
MATH 540: (MATH 234 or 375), (MATH 320, 340, 341, or 375), and (MATH 341, 375, 421, 467, or 521), or graduate/professional standing or member of the Pre-Masters Mathematics (Visiting International) Program
